File: scripts/portfolio_api.py
#!/usr/bin/env python3
"""
Programmatic JSON API for arbitrary Solana wallet portfolio reports.
"""

import argparse
import contextlib
import io
import json

try:
    from . import defi_tracker, portfolio_tracker
except ImportError:  # pragma: no cover
    import defi_tracker
    import portfolio_tracker


def _round_amount(value, digits=12):
    return round(float(value), digits)


def build_spot_portfolio(address: str) -> dict:
    assets = portfolio_tracker.get_assets_by_owner(address)
    positions = portfolio_tracker.parse_token_positions(assets)
    prices = portfolio_tracker.fetch_prices([position["mint"] for position in positions])

    tokens = []
    total_usd = 0.0
    for position in positions:
        price = float(prices.get(position["mint"], 0) or 0)
        usd_value = float(position["amount"]) * price
        total_usd += usd_value
        tokens.append(
            {
                "symbol": position["symbol"],
                "mint": position["mint"],
                "amount": _round_amount(position["amount"]),
                "price_usd": _round_amount(price),
                "value_usd": round(usd_value, 2),
                "token_type": position.get("token_type", ""),
            }
        )

    tokens.sort(key=lambda item: item["value_usd"], reverse=True)
    visible_tokens = [token for token in tokens if token["value_usd"] > 0.01]

    return {
        "total_usd": round(total_usd, 2),
        "token_count": len(tokens),
        "visible_token_count": len(visible_tokens),
        "hidden_token_count": len(tokens) - len(visible_tokens),
        "tokens": tokens,
    }


def build_defi_portfolio(address: str) -> dict:
    buffer = io.StringIO()
    with contextlib.redirect_stdout(buffer):
        positions = defi_tracker.detect_positions(address)
    positions.sort(key=lambda item: float(item.get("usd_value", 0) or 0), reverse=True)

    total_usd = 0.0
    protocols = {}
    normalized_positions = []
    for position in positions:
        usd_value = float(position.get("usd_value", 0) or 0)
        total_usd += usd_value

        normalized = {
            "protocol": position.get("protocol", ""),
            "type": position.get("type", ""),
            "account": position.get("account", ""),
            "url": position.get("url", ""),
            "value_usd": round(usd_value, 2),
            "details": position.get("details", {}),
        }
        normalized_positions.append(normalized)

        protocol_name = normalized["protocol"]
        protocol_bucket = protocols.setdefault(
            protocol_name,
            {
                "protocol": protocol_name,
                "total_usd": 0.0,
                "positions": [],
            },
        )
        protocol_bucket["total_usd"] += usd_value
        protocol_bucket["positions"].append(normalized)

    grouped_protocols = list(protocols.values())
    grouped_protocols.sort(key=lambda item: item["total_usd"], reverse=True)
    for protocol in grouped_protocols:
        protocol["total_usd"] = round(protocol["total_usd"], 2)

    return {
        "total_usd": round(total_usd, 2),
        "protocol_count": len(grouped_protocols),
        "position_count": len(normalized_positions),
        "protocols": grouped_protocols,
        "positions": normalized_positions,
    }


def get_wallet_portfolio_json(address: str) -> dict:
    """
    Return machine-readable spot + DeFi portfolio JSON for one Solana wallet.
    """
    spot = build_spot_portfolio(address)
    defi = build_defi_portfolio(address)
    return {
        "address": address,
        "spot": spot,
        "defi": defi,
        "totals": {
            "spot_usd": spot["total_usd"],
            "defi_usd": defi["total_usd"],
            "combined_usd": round(spot["total_usd"] + defi["total_usd"], 2),
        },
    }


def get_wallets_portfolio_json(addresses: list[str]) -> dict:
    """
    Return machine-readable combined portfolio JSON for multiple Solana wallets.
    """
    wallets = [get_wallet_portfolio_json(address) for address in addresses]
    spot_total = sum(wallet["totals"]["spot_usd"] for wallet in wallets)
    defi_total = sum(wallet["totals"]["defi_usd"] for wallet in wallets)

    return {
        "wallet_count": len(wallets),
        "wallets": wallets,
        "totals": {
            "spot_usd": round(spot_total, 2),
            "defi_usd": round(defi_total, 2),
            "combined_usd": round(spot_total + defi_total, 2),
        },
    }


def main():
    parser = argparse.ArgumentParser(description="Programmatic Solana portfolio JSON API")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--wallet", help="Single Solana wallet address")
    group.add_argument("--wallets", nargs="+", help="One or more Solana wallet addresses")
    args = parser.parse_args()

    if args.wallet:
        payload = get_wallet_portfolio_json(args.wallet)
    else:
        payload = get_wallets_portfolio_json(args.wallets)

    print(json.dumps(payload, indent=2))


if __name__ == "__main__":
    main()

File: scriptsTS/portfolio_api.ts
#!/usr/bin/env node

import { detectPositions } from "./defi_tracker.js";
import { fetchPrices, getAssetsByOwner, parseTokenPositions } from "./portfolio_tracker.js";
import { roundAmount } from "./lib/utils.js";
import type { WalletPortfolioJson, DefiPosition } from "./lib/types.js";

export async function buildSpotPortfolio(address: string): Promise<WalletPortfolioJson["spot"]> {
  const assets = await getAssetsByOwner(address);
  const positions = parseTokenPositions(assets);
  const prices = await fetchPrices(positions.map((position) => position.mint));

  const tokens = positions
    .map((position) => {
      const price = Number(prices[position.mint] || 0);
      const valueUsd = position.amount * price;
      return {
        symbol: position.symbol,
        mint: position.mint,
        amount: roundAmount(position.amount),
        price_usd: roundAmount(price),
        value_usd: Number(valueUsd.toFixed(2)),
        token_type: position.token_type,
      };
    })
    .sort((left, right) => right.value_usd - left.value_usd);

  const totalUsd = tokens.reduce((sum, token) => sum + token.value_usd, 0);
  const visibleTokens = tokens.filter((token) => token.value_usd > 0.01);

  return {
    total_usd: Number(totalUsd.toFixed(2)),
    token_count: tokens.length,
    visible_token_count: visibleTokens.length,
    hidden_token_count: tokens.length - visibleTokens.length,
    tokens,
  };
}

export async function buildDefiPortfolio(address: string): Promise<WalletPortfolioJson["defi"]> {
  const positions = await detectPositions(address);
  positions.sort((left, right) => Number(right.usd_value || 0) - Number(left.usd_value || 0));

  const normalized: DefiPosition[] = [];
  const protocolMap = new Map<string, { protocol: string; total_usd: number; positions: DefiPosition[] }>();
  let totalUsd = 0;

  for (const position of positions) {
    const usdValue = Number(position.usd_value || 0);
    totalUsd += usdValue;
    const normalizedPosition: DefiPosition = {
      protocol: position.protocol || "",
      type: position.type || "",
      account: position.account || "",
      url: position.url || "",
      value_usd: Number(usdValue.toFixed(2)),
      details: position.details || {},
    };
    normalized.push(normalizedPosition);

    const bucket =
      protocolMap.get(normalizedPosition.protocol) ||
      { protocol: normalizedPosition.protocol, total_usd: 0, positions: [] };
    bucket.total_usd += usdValue;
    bucket.positions.push(normalizedPosition);
    protocolMap.set(normalizedPosition.protocol, bucket);
  }

  const protocols = [...protocolMap.values()]
    .map((protocol) => ({ ...protocol, total_usd: Number(protocol.total_usd.toFixed(2)) }))
    .sort((left, right) => right.total_usd - left.total_usd);

  return {
    total_usd: Number(totalUsd.toFixed(2)),
    protocol_count: protocols.length,
    position_count: normalized.length,
    protocols,
    positions: normalized,
  };
}

export async function getWalletPortfolioJson(address: string): Promise<WalletPortfolioJson> {
  const spot = await buildSpotPortfolio(address);
  const defi = await buildDefiPortfolio(address);
  return {
    address,
    spot,
    defi,
    totals: {
      spot_usd: spot.total_usd,
      defi_usd: defi.total_usd,
      combined_usd: Number((spot.total_usd + defi.total_usd).toFixed(2)),
    },
  };
}

export async function getWalletsPortfolioJson(addresses: string[]): Promise<{
  wallet_count: number;
  wallets: WalletPortfolioJson[];
  totals: { spot_usd: number; defi_usd: number; combined_usd: number };
}> {
  const wallets = [];
  for (const address of addresses) {
    wallets.push(await getWalletPortfolioJson(address));
  }
  const spotTotal = wallets.reduce((sum, wallet) => sum + wallet.totals.spot_usd, 0);
  const defiTotal = wallets.reduce((sum, wallet) => sum + wallet.totals.defi_usd, 0);
  return {
    wallet_count: wallets.length,
    wallets,
    totals: {
      spot_usd: Number(spotTotal.toFixed(2)),
      defi_usd: Number(defiTotal.toFixed(2)),
      combined_usd: Number((spotTotal + defiTotal).toFixed(2)),
    },
  };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  let wallet: string | undefined;
  let wallets: string[] = [];

  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (arg === "--wallet") {
      wallet = args[index + 1];
      index += 1;
      continue;
    }
    if (arg === "--wallets") {
      wallets = args.slice(index + 1);
      break;
    }
  }

  try {
    const payload = wallet
      ? await getWalletPortfolioJson(wallet)
      : await getWalletsPortfolioJson(wallets);
    console.log(JSON.stringify(payload, null, 2));
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

if (process.argv[1] && process.argv[1].endsWith("portfolio_api.ts")) {
  void main();
}
